### DISCUSSION TITLE
Ways to stream data using python or Modbus prototype

### DISCUSSION THREAD

**Discussion by SanjeevKumarGoyal** [CURENT/andes#277] [Q&A] [answered]
Good Morning Dr. Cui,
I am looking for a way to streaming of data using python or any other way.
please suggest the possible ways to achieve it, your help in this regards would be much appreciable.

Regards,
Sanjeev

**Comment by jinningwang** [comment, marked as answer]
Hi Sanjeev,

This is Jinning. For ANDES we have integrated DiME for it to communicate data with other applications. For more info please refer to DiME: https://github.com/TheHashTableSlasher/dime

Regards,
Jinning

**Reply by SanjeevKumarGoyal** [reply]
Thanks Jinning,
let me go through it and ask you if come up with any questions.

Regards,
Sanjeev

**Comment by cuihantao** [comment]
DiME is the only one supported for now. You might need to write an adapter if you plan to use Modbus. I'm glad to help if you need help with locating data that is not covered in the examples or tutorial.

**Reply by SanjeevKumarGoyal** [reply]
Thanks Dr. Cui,
Please share the link to go through any documentation to use Modbus.
I have one question, which one is better to use either DiME or Modbus, and are they have same features and which one is much better to use and to understand.

Regards,
Sanjeev

**Comment by cuihantao** [comment]
The documents are the documentation and the example notebooks that you can find in the README.

DiME and Modbus are of two completely different categories. DiME is for data streaming simulation data to clients as other processes. Modbus is a communication protocol for embedded controllers.

Regards,
Hantao Cui

**Reply by SanjeevKumarGoyal** [reply]
Thanks Dr. Cui.

**Reply by SanjeevKumarGoyal** [reply]
hello Dr. Cui,

we want to connect **ANDES** to another program, that monitors the system conditions and then applies generation set points to some of the generators in ANDES.
we want the exchange of data to be in **MODBUS** to mimic the actual data transfer in the field.
Should we use **DiME** for this purpose or there is a way to use the MODBUS protocol?

Regards,
Sanjeev

**Comment by cuihantao** [comment]
If you need to use Modbus, then you will need to write a Modbus adapter for it. DiME will not be useful for this purpose.

Regards,
Hantao Cui

**Reply by SanjeevKumarGoyal** [reply]
hello Dr. Cui,

we want to use DiME to **stream the simulation data**, and on a periodic basis (e.g. every five minutes) to **change the generation set-points**. 
I would like to confirm whether DiME can accomplish this task, please confirm.

Regards,
Sanjeev

**Comment by cuihantao** [comment]
DiME can stream simulation data from ANDES; you will need to write a client to receive the data before processing. In terms of changing generator setpoints, I’m not sure if that feature is good or hadn’t been migrated. You might want to read the code in andes/io/streaming.py

Regards,
Hantao Cui

**Reply by SanjeevKumarGoyal** [reply]
Thanks Dr. Cui.

**Reply by SanjeevKumarGoyal** [reply]
Hello Dr. Cui,
i had gone through the Readme file and started basic implementation with that.
now we want to stream simulation data and gone through the streaming.py, tds.py and client.py files etc. and working with them.

the question is, if their is any kind of documentation about flow of all this stuff and working of all the functions used in streaming.py and other relevant files and way to verify as well, that would be helpful.

Regards,
Sanjeev

**Comment by cuihantao** [comment]
Hi,

I’m afraid I don’t have documentations other than the manual and those in the function dictating. For the undocumented parts, you will need to read the source code.

Regards,
Hantao Cui
